### PR TITLE
refactor mutex usage in the state package

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- The mutex was used on most of the AccountsDB functions.
  
## Proposed Changes
- Use the mutex only for critical sections. This will greatly increase the access time (most notably on API requests).

## Testing procedure
- Normal testing procedure

### CODE REVIEW

File: state/accountsDB.go [hunk 1]
```diff
@@ -542,20 +547,20 @@ func (adb *AccountsDB) saveAccountToTrie(accountHandler vmcommon.AccountHandler)
...
 func (adb *AccountsDB) RemoveAccount(address []byte) error {
-	adb.mutOp.Lock()
-	defer adb.mutOp.Unlock()
-
 	if len(address) == 0 {
 		return fmt.Errorf("%w in RemoveAccount", ErrNilAddress)
 	}
 
-	acnt, err := adb.getAccount(address)
+	adb.mutOp.Lock()
```
Comment (iulianpascalau) [inline state/accountsDB.go:560]: Please add a comment stating this is a critical section

**Review by AdoAdoAdo — DISMISSED** [review]

File: state/accountsDB.go [hunk 2]
```diff
@@ -270,26 +270,31 @@ func (adb *AccountsDB) GetCode(codeHash []byte) []byte {
...
 func (adb *AccountsDB) SaveAccount(account vmcommon.AccountHandler) error {
-	adb.mutOp.Lock()
-	defer adb.mutOp.Unlock()
-
 	if check.IfNil(account) {
 		return fmt.Errorf("%w in accountsDB SaveAccount", ErrNilAccountHandler)
 	}
 
-	oldAccount, err := adb.getAccount(account.AddressBytes())
+	adb.mutOp.Lock()
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/accountsDB.go:295]: Done.

File: state/accountsDB.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply state/accountsDB.go:562]: Done.

**Review by iulianpascalau — DISMISSED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

File: state/accountsDB_test.go [hunk 3]
```diff
@@ -2692,3 +2693,137 @@ func TestEmptyErrChanReturningHadContained(t *testing.T) {
...
+	numAccToLoad := 10
+	accounts := make([]vmcommon.AccountHandler, numAccToLoad)
+	for i := 0; i < numAccToLoad; i++ {
+		acc, err := adb.GetExistingAccount(addresses[i])
+		assert.Nil(t, err)
+		accounts[i] = acc
+	}
+	accountBytes, err := marshaller.Marshal(accounts[0])
+	assert.Nil(t, err)
+	counter := atomicFlag.Counter{}
```
Comment (iulianpascalau) [inline state/accountsDB_test.go:2727]: in this case we can safely remove counter and put wg.Done after the switch

File: state/accountsDB_test.go [hunk 4]
```diff
@@ -2692,3 +2693,137 @@ func TestEmptyErrChanReturningHadContained(t *testing.T) {
 		})
 	})
 }
+
+func TestAccountsDb_Concurrent(t *testing.T) {
+	_, adb := getDefaultTrieAndAccountsDb()
+
+	numAccounts := 100000
```
Comment (iulianpascalau) [inline state/accountsDB_test.go:2700]: can put this to 1000 as the race detector can handle it

File: state/accountsDB_test.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply state/accountsDB_test.go:2727]: Removed.

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Pull request merged**